Add WASM bindings for `moyo`

janosh opened the pull request:
closes #103

lan496 commented:
Cool! Just out of curiosity, how is this WASM binding used in matterviz side? 

Also, can I set you as [CODEOWENERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) under `moyo-wasm`?

janosh commented:
the first pass at a new symmetry demo page using `moyo`'s WASM bindings is here branch https://github.com/janosh/matterviz/compare/main...symmetry

i'm installing my local [`moyo` clone to test this in the dev server](https://github.com/janosh/matterviz/compare/main...symmetry#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R32). needs a `moyo-wasm` NPM release after this PR is merged before i can deploy to GitHub Pages

> Also, can I set you as [CODEOWENERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) under moyo-wasm?

of course, thank you!

janosh commented:
i changed the layout and added some more information. if you have any thoughts on what other info would be good to show, i'm all ears!

https://github.com/user-attachments/assets/5f2cd17f-7728-4e2c-a6e3-328c650bb829

joonatan-altrove commented:
Some quick thoughts from a crystallographer: 
- super nice visuals! 
- Writing the Hermann-Mauguin notation for the space group would be very useful along the space group number (e.g. Fd-3m for space group 227). The number itself is hard to remember, but the HM notation contains all relevant information to get a thought on the space group itself
- I think the updated layout is much clearer
- it would probably be worth it to also list the non-symmetric sites in the list (I don't see the Mg atoms in Ac4Mg2, for example). Is that a bug or intended feature?
- one thing that could be useful is a way to visualise the symmetry elements. For example, visualise the mirror planes / rotational axes. Might be overkill to implement.
- how are you dealing with disordered / fractionally occupied sites?

janosh commented:
> how are you dealing with disordered / fractionally occupied sites?

not well https://github.com/spglib/moyo/issues/110 😅

joonatan-altrove commented:
ahh😅
disordered sites are a bit tricky in general. Maybe they can be visualised as an atom-like object (a composite atom that could be multiple different elements at certain ratios), or maybe just visualised as the most likely atom + a note about occupancy

lan496 commented on .github/workflows/publish-wasm.yml:
```diff
@@ -0,0 +1,31 @@
+      - name: Publish to npm
+        working-directory: moyo-wasm/pkg
+        env:
+          NODE_AUTH_TOKEN: ${{ secrets.NPM_TOKEN }}
```
Should I set some token for npm?

janosh commented:
@lan496 i had a tough time getting strongly typed WASM bindings via an auto-generated TypeScript module declarations file (`.d.ts`).  [`moyo_wasm.d.ts`](https://github.com/user-attachments/files/21936256/moyo_wasm.d.ts) is the current best one i managed so far with the code pushed in 0246eb5. it enables a nicer developer experience on the JS side from auto-completions and type errors when using attributes incorrectly:

<img width="1308" height="700" alt="Screenshot 2025-08-22 at 2 02 07 PM" src="https://github.com/user-attachments/assets/6890c212-1849-4ba2-a541-4fa99e9eeb4d" />

i don't really know what i'm doing when it comes to tools like [`tsify`](https://github.com/madonoharu/tsify) or [`ts-rs`](https://github.com/Aleph-Alpha/ts-rs) so it's possible that the current solution is inelegant. i'll just post for later reference what i tried/considered:

- JSON-only + consumer-defined TS types (rejected)
  - What: `analyze_cell(...) -> JsValue` via `serde_wasm_bindgen`; leave all type definitions to the WASM consumer (terrible DX)
  - Pros: Smallest WASM surface; zero glue in Rust.
  - Cons: Type drift risk (consumer must mirror Rust types manually); lots of maintenance over time.

- `wasm-bindgen` typed-array API (analyze_simple_cell)
  - What: Accept `Float64Array/Int32Array`, return JS object; no JSON parsing in Rust.
  - Pros: Efficient data path, fully JS-native arrays.
  - Cons: Dual API surface; manual marshalling; still no authoritative .d.ts from Rust; more moving parts.
  - Status: Rejected for extra API + maintenance.

- `ts-rs` (derive TS from core `moyo` types via `cfg_attr` + `build.rs`)
  - What: Annotate core structs (`Lattice`, `Cell`, `Operation`, `MoyoDataset`) with `#[cfg_attr(feature="ts", ...)]` and generate `.d.ts` in a build script.
  - Pros: Perfect single source of truth from the real structs; no WASM DTOs.
  - Cons:
    - Required pervasive annotations across core crate (matrices, vectors, enums).
    - Workspace build-dep and feature-gating complexity (build.rs had to import `moyo` with `ts` feature; had cfg issues that i couldn't resolve).
    - Publishing needed an extra “ts” build and copying generated .d.ts into `pkg/` before `wasm-pack`.
    - Still needed field-level TS overrides for `nalgebra` types (Matrix3, Vector3), so boilerplate didn’t fully disappear. maybe the best solution in principle but I didn't manage

- `tsify` boundary layer (the solution pushed in 0246eb5)
  - What: Small, WASM-facing structs in `moyo-wasm/lib.rs` that derive `Tsify` and are `wasm-bindgen` compatible; flatten `nalgebra` types; return that from `analyze_cell`.
  - Pros:
    - `wasm-pack`/`wasm-bindgen` emits a `.d.ts` directly inside `pkg/` with no extra build steps.
    - Types versioned with the WASM package; no consumer drift.
    - Clear boundary; minimal impact on the core crate.
    - Single export (`analyze_cell`) yields full dataset; good DX.
    - ships authoritative TypeScript definitions with the npm package without brittle workspace feature gates, extra CI steps, and the glue code is contained `moyo-wasm/lib.rs`
  - Cons: Some boilerplate in `moyo-wasm/lib.rs` to flatten matrices/vectors and map core types to exports (30 lines became 180 lines)

janosh commented:
> it would probably be worth it to also list the non-symmetric sites in the list (I don't see the Mg atoms in Ac4Mg2, for example). Is that a bug or intended feature?

@joonatan-altrove good catch, that's definitely a bug! fixed (only partially, i think) in https://github.com/janosh/matterviz/commit/eac0157082195a0c59e375feea44df1a51d668cb. will take another look tomorrow but doesn't hold up this PR

lan496 commented on moyo/src/lib.rs:
```diff
@@ -241,6 +243,7 @@ impl MoyoDataset {
             // Space-group type
             number: space_group.number,
             hall_number: space_group.hall_number,
+            hm_symbol: hall_symbol.hm_short.to_string(),
```
The current `MoyoDataset` takes one extreme by removing every redundant entry (for example, hm_short is not included because it can be derived from `number`). The other extreme would be to include all secondary entries. 
I was initially unsure how to balance the two extremes and choose secondary information (https://github.com/spglib/moyo/issues/77#issuecomment-2684049654). But on second thought, I think it is reasonable to include `hm_short`. It is frequently needed for quick access, as you pointed out, and [spglib dataset](https://spglib.readthedocs.io/en/stable/dataset.html) offers hall_symbol for convenience
